pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1337 @@
+					Self::deposit_event(Event::TradableStateUpdated { asset_id, state });
+
+					Ok(())
+				})
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Protocol account address
+	fn protocol_account() -> T::AccountId {
+		PalletId(*b"omnipool").into_account()
+	}
+
+	/// Convert asset weight cap to FixedU128
+	fn asset_weight_cap() -> Price {
+		let fee = T::AssetWeightCap::get();
+		match fee {
+			(_, 0) => FixedU128::zero(),
+			(a, b) => FixedU128::from((a, b)),
+		}
+	}
+
+	/// Retrieve stable asset detail from the pool.
+	/// Return NoStableCoinInPool if stable asset is not yet in the pool.
+	fn stable_asset() -> Result<(Balance, Balance), DispatchError> {
+		let stable_asset = <Assets<T>>::get(T::StableCoinAssetId::get()).ok_or(Error::<T>::NoStableAssetInPool)?;
+		let stable_reserve = T::Currency::free_balance(T::StableCoinAssetId::get(), &Self::protocol_account());
+		Ok((stable_reserve, stable_asset.hub_reserve))
+	}
+
+	/// Retrieve state of asset from the pool and its pool balance
+	fn load_asset_state(asset_id: T::AssetId) -> Result<AssetReserveState<Balance>, DispatchError> {
+		let state = <Assets<T>>::get(asset_id).ok_or(Error::<T>::AssetNotFound)?;
+		let reserve = T::Currency::free_balance(asset_id, &Self::protocol_account());
+		Ok((state, reserve).into())
+	}
+
+	/// Set new state of asset.
+	/// This converts the new state into correct state type ) by removing the reserve)
+	fn set_asset_state(asset_id: T::AssetId, new_state: AssetReserveState<Balance>) {
+		<Assets<T>>::insert(asset_id, Into::<AssetState<Balance>>::into(new_state));
+	}
+
+	/// Generate an nft instance id and mint NFT into the class and instance.
+	#[require_transactional]
+	fn create_and_mint_position_instance(owner: &T::AccountId) -> Result<T::PositionInstanceId, DispatchError> {
+		<PositionInstanceSequencer<T>>::try_mutate(|current_value| -> Result<T::PositionInstanceId, DispatchError> {
+			let next_position_id = *current_value;
+
+			T::NFTHandler::mint_into(&T::NFTClassId::get(), &next_position_id, owner)?;
+
+			*current_value = current_value
+				.checked_add(&T::PositionInstanceId::one())
+				.ok_or(ArithmeticError::Overflow)?;
+
+			Ok(next_position_id)
+		})
+	}
+
+	/// Update Hub asset side of HDX subpool annd add given amount to hub_asset_reserve
+	fn update_hdx_subpool_hub_asset(hub_asset_amount: Balance) -> DispatchResult {
+		if hub_asset_amount > Balance::zero() {
+			let mut native_subpool = Assets::<T>::get(T::NativeAssetId::get()).ok_or(Error::<T>::AssetNotFound)?;
+			native_subpool.hub_reserve = native_subpool
+				.hub_reserve
+				.checked_add(hub_asset_amount)
+				.ok_or(ArithmeticError::Overflow)?;
+			<Assets<T>>::insert(T::NativeAssetId::get(), native_subpool);
+		}
+		Ok(())
+	}
+
+	/// Update total hub asset liquidity and write new value to storage.
+	/// Update total issueance if AdjustSupply is specified.
+	#[require_transactional]
```

Comment:
We need to use `require_transactional` if they both:
+ mutate storage
+ and are also fallible

And could thus lead to a partial update, i.e. the first write succeeds, but the second write is skipped because the function errors out.

Reply:
@apopiak got it now, thank you, Alex!